### DISCUSSION TITLE
[RFC] Deterministic Draft Filter: Pluggable Domain Validator Framework for MTP Speculative Decoding

### DISCUSSION THREAD

**Discussion by SamD** [ggml-org/llama.cpp#25219] [Ideas]
*Apologies for the longer post, but it's somewhat necessary to cover the full scope of what's being proposed here.*
 
### Background
 
Running coding and tool-calling agents on commodity hardware (specifically a
BeeLink N100 mini PC with 4 efficiency cores) was the starting point. llama.cpp
was the natural choice: lightweight, full control over KV quantization,
parallelism, thread and batch sizing. MoE and MTP both helped, and it was the
MTP path that gave me the idea.
 
MTP works by generating draft tokens from auxiliary heads and verifying them
against the target model. That verification step is a joinpoint: a natural place to intercept. If the domain is well understood and has a well-defined
grammar, a deterministic validator at that joinpoint can make a more accurate
and faster determination about token validity than probabilistic logit sampling
against the target model ever could.
 
The initial goal was a full coding replacement. Tree-sitter supports a large
number of languages and is freely available, so it seemed like the right tool.
It turned out to be more cumbersome to integrate with the MTP pipeline than
expected, and it exposed a more fundamental problem: tree-sitter can verify
syntax, but if a draft head produces tokens that are syntactically correct but
semantically wrong (a plausible but nonexistent function name, a variable out
of scope) without introducing a compiler. Disappointing, but
it clarified something important.
 
Code generation is one domain. There are many others: regulated environments,
private schemas, financial formats, legal citation rules, where the domain
owner knows the rules better than any model does, and where rigid constraints
are not a limitation but a requirement. Tree-sitter is not generic enough for
that. XGrammar is. Define your own grammar, get your own bitmask-based
validator, plug it in. The architecture needed to support that.
 
The N100 result (0.12 tps to 3.30 tps) is meaningful but not the end goal.
A capable edge agent needs model capacity, which remains the harder constraint.
This work makes the integration point available. What gets built on it is a
separate problem.
 
### Proposal
 
*A working PoC exists for everything described below.*
 
A pluggable framework, based on the MTP pattern already in place, that
intercepts draft tokens at the joinpoint where target verification would occur
and routes them through a domain-specific validator loaded at runtime via
`dlopen`/`dlsym`. With a good quality model producing long valid runs, the
plugin passes draft batches through directly. Far fewer decode operations
against the target model, faster TPS. With a weaker model or aggressive
quantization producing garbage drafts, the plugin catches them early via
bitmask, preventing worse outcomes.
 
The plugin contract is the contribution. The XGrammar reference implementation
is one domain among many.
 
### The interception point
 
After MTP draft tokens are generated, the normal path is target model
verification via `common_sampler_sample_and_accept_n`. With
`--det-draft-accept-all`, the plugin takes ownership of that decision entirely:
 
```c
if (common_speculative_get_det_accept_all(spec)) {
    for (size_t i = 0; i < draft.size(); i++) {
        common_sampler_accept(smpl, draft[i], true);
        result.push_back(draft[i]);
    }
```
 
The target model's verification logic is not consulted for plugin-accepted
tokens. Domain logic that is more authoritative than probabilistic logit
sampling belongs in the plugin.
 
The speedup is not from skipping the target forward pass. The target model
still runs and its KV cache still advances. Its logits are only used to sample
the next token, not to verify the draft. The gain is from high draft acceptance
rates instead of near-total rejection.
 
### Core cleanliness
 
The llama.cpp core is not polluted. No third-party dependencies have been
introduced into the core tree. No implementation details from any plugin leak
into core. The core changes are limited to the common interface -- shared data
structures, headers, three CLI flags, and the dynamic loader. The loader calls
into a plugin only if one has been explicitly loaded via `--det-draft-model`.
Without that flag, no plugin code runs and the code path is identical to
unmodified llama.cpp.
 
XGrammar, tree-sitter, or any other domain-specific library lives entirely
in the plugin. Core knows nothing about them.
 
### Plugin architecture
 
```
PLUGIN AUTHOR          DISTRIBUTION           END USER
─────────────          ────────────           ────────
plugin.cpp
#include plugin.h
implements contract
      |
      v
deterministic-        ships .so         downloads .so
draft.so         ─────────────────►           |
                                              v
                                    llama.cpp
                                    --det-draft-model ./plugin.so
                                         |
                                    libdeterministic_draft_spec.so
                                    dlopen / dlsym
                                    deterministic_draft_filter_draft
                                    deterministic_draft_commit
                                    deterministic_draft_rollback
                                    deterministic_draft_reset
                                    deterministic_draft_destroy
                                         |
                                    common/speculative.cpp
                                    feeds draft tokens to plugin
                                    applies filter result
```
 
**Plugin author**: writes a domain-specific validator, includes
`deterministic_draft_plugin.h`, implements the contract, compiles to a `.so`.
The only party that needs the headers. The XGrammar reference implementation
is one example; a regulated organisation's private validator is another.
 
**Distribution**: ships the `.so`. Open source, private corporate artifact,
or anything in between. llama.cpp carries no opinion on distribution.
 
**End user**: passes `--det-draft-model ./plugin.so` and runs llama.cpp.
Never sees a header. `libdeterministic_draft_spec.so` handles `dlopen`/`dlsym`
at runtime and calls into the plugin through `common/speculative.cpp`.
 
### Plugin contract
 
The contract is capability-based. Plugins declare what they support via
`deterministic_draft_get_capabilities()` and the host degrades gracefully for
missing capabilities.
 
**CAPABILITY_BITMASK**: the plugin fills a bitmask of valid token IDs before
each step. `deterministic_draft_filter_draft()` filters a batch of draft tokens
against the bitmask, commits valid tokens to grammar state, and stops at the
first invalid token. This is the primary validation mechanism in the reference
implementation.
 
**CAPABILITY_JUMP_FORWARD**: the plugin returns strings uniquely determined
by the current grammar state, allowing deterministic sequences to be skipped
without model sampling. Present in the plugin contract and implemented in the
reference implementation. Not yet wired into `common/speculative.cpp` -
integration into the MTP draft stream is non-trivial and deferred to follow-up
work.
 
**Rollback**: `deterministic_draft_rollback()` undoes the last N commit calls
for a given slot, keeping grammar state consistent with what was actually
emitted when the target model accepts fewer tokens than the plugin already
committed. Required for correct behaviour in default mode.
 
**Multi-slot isolation**: each inference slot carries a `slot_id`. Plugins
maintain per-slot state internally. The server's concurrent inference slots are
correctly isolated.
 
### Core changes to llama.cpp
 
- Shared data structures and headers
- Three CLI flags
- `external/deterministic_draft_spec.cpp` - plugin loader (dlopen/dlsym)
- `external/include/llama_deterministic_draft.h` - consumer API header
- `external/include/deterministic_draft_plugin.h` - plugin contract header
- `common/speculative.{h,cpp}` - pipeline integration, conditional bypass of
  `common_sampler_sample_and_accept_n` when both flags are set
Without `--det-draft-model` the code path is identical to unmodified llama.cpp.
With `--det-draft-model` but without `--det-draft-accept-all`,
`common_sampler_sample_and_accept_n` runs as normal. Only when both flags are
set does core bypass target verification. The plugin has no access to sampling
routines.
 
Build with `-DDETERMINISTIC_SPEC_ENABLED=ON`. Without this flag core is
unchanged.
 
### Relationship to MTP
 
MTP is reused as-is. Enabling `--det-draft-model` auto-enables
`--spec-type draft-mtp`. If the model lacks MTP auxiliary heads
(`n_layer_nextn == 0`), llama.cpp fails with the standard MTP error.
 
### Flags
 
| Flag | Description |
|---|---|
| `--det-draft-model FNAME` | Path to plugin (.so/.dylib/.dll). Auto-enables `--spec-type draft-mtp`. |
| `--det-draft-n-max N` | Max draft tokens per step. When >0, also sets `--spec-draft-n-max` to N. |
| `--det-draft-accept-all` | Plugin takes ownership of verification. Default: false. |
| `--lang LANGUAGE` | Select grammar by language name (benchmark tool). |
 
`--det-draft-accept-all` is appropriate only for single-language code-only
generation or domains where the plugin can be fully trusted as the authority.
The plugin validates structural correctness only; semantic correctness remains
the caller's responsibility.
 
### Reference implementation
 
The PoC uses XGrammar for grammar-constrained decoding with bitmask-based
token filtering. Grammars are plain `.gbnf` files bundled alongside the `.so`
and resolved at runtime by language name. Define your own grammar file, get
your own domain-specific validator.
 
The filter validates structural correctness only, not semantic correctness. A
syntactically valid token sequence may still be semantically wrong. For domains
where structural validity is sufficient, or where the domain owner defines
what validity means, the plugin is the right authority.
 
### Benchmarks
 
These are illustrative results from one model and prompt configuration. Speedup
is dependent on hardware, model quality, quant, prompt, grammar, and n_max.
Treat them as indicative, not guaranteed.
 
Model: unsloth/Qwen3.5-2B-MTP-GGUF (Q4_K_M). All runs:
`--det-draft-n-max 100 --det-draft-accept-all --lang c`.
Full methodology in the fork docs.
 
#### Summary (fib prompt, 200 tokens, 3 runs)
 
| Hardware | Baseline TPS | Treatment TPS | Speedup |
|---|---|---|---|
| RTX 4070 (CUDA) | 21.94 | 27.30 | 1.24x |
| GTi15 Arc PRO B70 (oneAPI) | 19.47 | 130.88 | 6.72x |
| N100 (CPU only) | 0.12 | 3.30 | 28.1x |
 
The RTX 4070 result reflects a short single-line prompt. Speedup scales with
prompt complexity and draft size; with longer real code prompts the numbers
improve considerably. Speedup is generally higher on slower hardware because
baseline verification waste is proportionally larger.
 
#### N100 detail (fib prompt, single run)
 
| Metric | Baseline (MTP) | Treatment (MTP+DET) | Improvement |
|---|---|---|---|
| throughput tps | 0.12 | 3.30 | +2712% |
| accept rate | 0.4% | 100.0% | +99.6pp |
| drafted (pre) | 14333 | 200 | 72x fewer |
| drafted (post) | 14333 | 133 | 108x fewer |
| det truncated | N/A | 67 | - |
 
On the N100, 99.6% of draft tokens were being rejected before the plugin. The
plugin eliminates that waste. 0.12 tps to 3.30 tps: unusable to interactive.
 

### Fork

[deterministic fork](https://github.com/SamD/llama-cpp-deterministic-fork)

Documentation:
- [README](https://github.com/SamD/llama-cpp-deterministic-fork/blob/deterministic-draft-spec/deterministic-draft-model-poc/README.md)
- [Quick Start](https://github.com/SamD/llama-cpp-deterministic-fork/blob/deterministic-draft-spec/deterministic-draft-model-poc/docs/quick-start.md) - build and run instructions
- [Benchmark Overview](https://github.com/SamD/llama-cpp-deterministic-fork/blob/deterministic-draft-spec/deterministic-draft-model-poc/docs/benchmark-overview.md) - multi-hardware results

**Comment by SamD** [comment]
@am17an , given your experience with MTP (#22673/#23269), would love your take on this proposal, especially whether the plugin extension point makes sense architecturally.

Thanks

**Reply by am17an** [reply]
At a cursory glance your proposal, it seems like you might benefit from training a DSpark head on your domain specific knowledge

**Reply by SamD** [reply]
Thank you for the fast response, wasn't expecting it :)

From what I gather the DSpark proposal is more about improving the accuracy of the draft heads.
This would certainly be improvement and if it is implemented it would enhance my proposal even further as I would see them being more complementary vs overlap.

I know my post was long but the gist of is this
1. Core is clean a shared contract 
2. The plugin itself is domain specific, I provided a working PoC with xgrammar and gbnf but this is only an example.
 Originally I wanted to do a full blown solution but discovered later without a compiler it wouldn't work since all we can ensure is the syntax. But there are many domains, business, finance, law, defense ... anything that requires script compliance cannot risk a probabilistic outcome
3. It leverages MTP and follows the pattern, 

In short, a clean plugin following an existing pattern (MTP) that provides overriding the probabilistic verification of the main llm with that deterministic open, closed implementation where it makes sense where the domain is known, a grammar is defined and absolutely necessary as in the case of regulatory environments.

**Comment by SamD** [comment]
@am17an, wanted to close the loop on this rather than let it go quiet.

One clarification on framing first: the original motivation here was throughput, eliminating the main LLM verification overhead on weak drafts, not the domain-validation angle. A simple filter alone would already be faster, since it cuts the wasted draft/verify cycles that dominate on weak hardware (the N100 numbers are really this effect). The deterministic, per-domain validation piece (grammar-defined correctness, no probabilistic outcome) is a bonus this architecture happens to unlock, useful in specific cases where that's a hard requirement, but it wasn't the starting goal.

On the DSpark-head point specifically: I still see the two as complementary rather than overlapping. DSpark-style training improves the *accuracy* of the draft heads themselves (fewer bad drafts to begin with), which helps the throughput goal directly. The plugin sits at the verification joinpoint regardless of draft quality, so better draft heads just mean less for it to catch, good for both approaches, and the deterministic-guarantee use case remains available on top for domains that need it.

Given there's a working fork with docs and multi-hardware benchmarks already, I'm considering opening this as a draft PR so the actual diff is easier to review than the discussion prose. Would that be useful, or is there a preferred process for something at this scope (e.g. splitting into smaller PRs, an issue first, etc.)? Also happy to loop in whoever else might want visibility on something at this scope, let me know if that's useful.

Thanks again for engaging with this, appreciate the sanity check.

**Reply by am17an** [reply]
> On the DSpark-head point specifically: I still see the two as complementary rather than overlapping. DSpark-style training improves the accuracy of the draft heads themselves (fewer bad drafts to begin with), which helps the throughput goal directly

I would suggest you to actually read the paper than relying on LLM outputs. Not only does it increase the accurary of the draft heads but it also drops tokens to verify based on the confidence thresholds.

**Reply by SamD** [reply]
@am17an, thanks for pointing me to this, I hadn't read the DSpark paper yet and will go through it properly before saying anything more definitive.

One thing I did want to raise in the meantime, not as a counterpoint, just a framing distinction I think matters: however good a confidence estimate gets, it's still a probability, calibrated or not. A domain specific grammar doesn't estimate correctness, it verifies it deterministically. For most domains that difference doesn't matter much and DSpark's approach is probably the better general-purpose tool. But for the narrow set of domains this RFC is actually aimed at, regulated formats (defense , finance ... etc.), fixed schemas, anywhere a structurally invalid output isn't tolerable regardless of confidence, that gap is the whole point, and it likely comes with a speed advantage too, since a grammar check on known-valid tokens should be cheaper than scoring and calibrating a confidence estimate.

So I'd frame it less as competing solutions and more as DSpark for the general case, this for the narrow, zero-tolerance case where it applies which is more niche but an important one to have available if needed. Appreciate you pushing me to actually engage with the paper rather than assume, will follow up once I've read it properly.

**Reply by SamD** [reply]
@am17an , I've gone through the DSpark paper (https://arxiv.org/pdf/2607.05147) more thoroughly and it's definitely helped solidify my understanding of the MTP pipeline.

One thing that stood out was how the standard autoregressive approach creates significant overhead. Where each draft token requires a full forward pass through the model, duplicating n transformations just for drafting before verification even begins. The DSpark parallel drafters mitigate this by collapsing all n tokens into a single pass, which is where most of the latency gains come from. The tradeoff is collisions when different phrases are possible, e.g. 'of course' and 'no problem' turning into 'of problem', a problem noted back in 2018 when parallel decoding was first tried. To fix this, a lightweight sequential head injects inter-token dependencies without adding meaningful latency. The confidence-scheduled verification then drops low-confidence suffix tokens before they reach the target model, freeing up batch capacity.

From my reading, my RFC aligns with DSpark's improvements in what gets drafted and how verification is scheduled at the system level, while the plugin operates independently at the verification joinpoint (borrowing the AOP term, since that's really what this is, intercepting a well-defined point in the pipeline) regardless of draft quality. They're tackling different layers of the same pipeline. For general use cases, DSpark training seems like the better option, but for regulated domains where structural validity is critical, and where a grammar check on known-valid tokens should be cheaper than scoring + calibrating confidence estimates, the deterministic path remains the right choice.

I think the framing as complementary rather than competing holds up. Thanks for pointing me to the paper, it clarified a lot about where the overhead actually lives in the pipeline.

**Comment by SamD** [comment]
https://github.com/ggml-org/llama.cpp/pull/26551

**Comment by SamD** [comment]
The PR has been updated to include changes introduced by the latest rebase from upstream which pulled in significant changes for
eagle3/dspark/dflash.
